File: Java-Kotlin/src/main/java/nowcoder/nc128.java
package nowcoder;

public class nc128 {

    // see blue book
    public long maxWater(int[] arr) {
        // write code here
        if (!isValid(arr)) {
            return 0;
        }

        int left = 0;
        int right = arr.length - 1;
        long lMax = arr[left];
        long rMax = arr[right];
        long ans = 0;
        while (left <= right) {
            lMax = Math.max(lMax, arr[left]);
            rMax = Math.max(rMax, arr[right]);
            if (lMax < rMax) {
                ans += lMax - arr[left];
                left++;
            } else {
                ans += rMax - arr[right];
                right--;
            }
        }
        return ans;
    }

    private static boolean isValid(int[] arr) {
        if (arr == null || arr.length == 0) {
            return false;
        }
        return true;
    }
}

File: Go-Solutions/nowcoder/nc128.go
package nowcoder

// see blue book
func maxWater128(arr []int) int64 {
	// write code here
	if len(arr) == 0 {
		return 0
	}

	left, right := 0, len(arr)-1
	lMax, rMax := arr[left], arr[right]
	ans := 0
	for left <= right {
		lMax = max128(lMax, arr[left])
		rMax = max128(rMax, arr[right])

		if lMax < rMax {
			ans += lMax - arr[left]
			left++
		} else {
			ans += rMax - arr[right]
			right--
		}
	}
	return int64(ans)
}

func max128(a, b int) int {
	if a > b {
		return a
	}
	return b
}
